src/compiler.jl

```diff
@@ -2204,6 +2204,59 @@ GPUCompiler.runtime_slug(job::CompilerJob{EnzymeTarget}) = "enzyme"
 GPUCompiler.get_interpreter(job::CompilerJob{<:Any,<:AbstractEnzymeCompilerParams}) =
     EnzymeInterpeter(GPUCompiler.ci_cache(job), GPUCompiler.method_table(job), job.source.world)
 
+using JET.JETInterface
+import JET: report_call, get_reports
+
+struct EnzymeAnalyzer <: AbstractAnalyzer
+	state::AnalyzerState
+	cache_key::UInt
+end
+function EnzymeAnalyzer(; jetconfigs...)
+	return EnzymeAnalyzer(
+		AnalyzerState(; jetconfigs...),
+		__ANALYZER_CACHE_KEY[] += 1,
+		)
+end
+const __ANALYZER_CACHE_KEY = Ref{UInt}(0)
+
+JETInterface.AnalyzerState(analyzer::EnzymeAnalyzer) = analyzer.state
+function JETInterface.AbstractAnalyzer(analyzer::EnzymeAnalyzer, state::AnalyzerState)
+    return EnzymeAnalyzer(
+        state,
+		analyzer.cache_key,
+        )
+end
+JETInterface.ReportPass(analyzer::EnzymeAnalyzer) = EnzymeAnalysisPass()
+JETInterface.get_cache_key(analyzer::EnzymeAnalyzer) = analyzer.cache_key
+
+function Core.Compiler.finish(frame::InferenceState, analyzer::EnzymeAnalyzer)
+	ReportPass(analyzer)(UnionResultReport, analyzer, frame)
+    return Base.@invoke Core.Compiler.finish(frame::InferenceState, analyzer::AbstractAnalyzer)
+end
+
+struct EnzymeAnalysisPass <: ReportPass end
+
+@reportdef struct UnionResultReport <: InferenceErrorReport
+	@nospecialize(bestguess) # ::Type
+end
+JETInterface.get_msg(::Type{UnionResultReport}, _, @nospecialize(bestguess)) = "Union result (`$bestguess`) detected"
+function (::EnzymeAnalysisPass)(::Type{UnionResultReport}, analyzer::EnzymeAnalyzer, frame::InferenceState)
+    if isa(frame.bestguess, Union)
+        add_new_report!(analyzer, frame.result, UnionResultReport(frame.linfo, frame.bestguess))
+        return true
+    end
+    return false
+end
+
+function GPUCompiler.ci_cache_populate(interp::EnzymeInterpeter, cache, mt, mi, min_world, max_world)
+	jetresult = report_call(mi.specTypes; analyzer=EnzymeAnalyzer)
```

Comment:
Even beyond that, we should be able to handle unions if the two datatypes are pointers, yes?

I'm inclined to have this be something that runs optionally right now (and perhaps we get it to properly match the "what is handled" as time goes on).

Reply:
A second avenue would be instead of:

```
Illegal updateAnalysis prev:{[-1]:Float@double} new: {[-1]:Integer}
val:   %4 = bitcast double %1 to i64, !dbg !5 origin=  %.sroa.02.0 = select i1 %.not, i64 0, i64 %4, !dbg !5
julia: /workspace/srcdir/Enzyme/enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:643: void TypeAnalyzer::updateAnalysis(llvm::Value*, TypeTree, llvm::Value*): Assertion `0 && "Performed illegal updateAnalysis"' failed.

signal (6): Aborted
in expression starting at REPL[5]:1
__pthread_kill_implementation at /usr/bin/../lib/libc.so.6 (unknown line)
raise at /usr/bin/../lib/libc.so.6 (unknown line)
abort at /usr/bin/../lib/libc.so.6 (unknown line)
__assert_fail_base.cold at /usr/bin/../lib/libc.so.6 (unknown line)
__assert_fail at /usr/bin/../lib/libc.so.6 (unknown line)
updateAnalysis at /workspace/srcdir/Enzyme/enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:643
visitSelectInst at /workspace/srcdir/Enzyme/enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:1777
run at /workspace/srcdir/Enzyme/enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:1087
```

Hitting an assertion error to firstly call `julia_error` and secondly to pass the module and instruction as well so that we can decode the `!dbg !5` chain and report a better error.

(Even if we don't call julia_error we should print a stack trace on assertion that is in user code not just system code.)

Reply:
> How bad is the overhead here? Is this causing an inference using JETAbstractInterpreter?

The analysis uses JET's own `AbstractInterpreter` (not integrated with Enzyme's `AbstractInterpreter`), and it runs fresh analysis. So there would be duplicated work in Julia-level inference.

In terms of correctness, this non-composed approach is okay for now since `EnzymeInterpreter` doesn't change anything about abstract interpretation (it does tweak the optimization by chaning inlining heuristics, but it doesn't matter for JET's type-level analysis).



> I'm inclined to have this be something that runs optionally right now (and perhaps we get it to properly match the "what is handled" as time goes on).

Yeah, sure. Maybe it would be reasonable to setup an optional keyword for `autodiff(... ; check::Bool=false)`?

Reply:
> Yeah, sure. Maybe it would be reasonable to setup an optional keyword for autodiff(... ; check::Bool=false)?

I think I would prefer an `@report_enzyme` or something.

Reply:
Done: please see docstrings for the analysis APIs.
https://github.com/EnzymeAD/Enzyme.jl/pull/261/files?diff=unified&w=1